main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FrequencyTimeInOrderValidator.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 Google LLC, MobilityData IO
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.validator;
+
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.notice.StartAndEndTimeOutOfOrderNotice;
+import org.mobilitydata.gtfsvalidator.table.GtfsFrequency;
+import org.mobilitydata.gtfsvalidator.type.GtfsTime;
+
+import static org.mobilitydata.gtfsvalidator.table.GtfsFrequencyTableLoader.FILENAME;
+
+/**
+ * Validates `frequencies.start_time` is before or equal to `frequencies.end_time` for a single entity.
+ * <p>
+ * Generated notices:
+ * * StartAndEndTimeOutOfOrder
+ */
+@GtfsValidator
+public class FrequencyTimeInOrderValidator extends SingleEntityValidator<GtfsFrequency> {
+    @Override
+    public void validate(GtfsFrequency frequency, NoticeContainer noticeContainer) {
+        // startTime and endTime are assumed to be not null: therefore, we do not check hasStartTime and hasEndTime.
+        GtfsTime startTime = frequency.startTime();
+        GtfsTime endTime = frequency.endTime();
+        if (startTime.isAfter(endTime)) {
```

Comment:
Sorry, I missed this when I looked at it previously - so what is the behavior when these values are `null` in a real dataset? Do we stop earlier in the validation process, or will this line produce an NPE?

If the `validate()` method executes and doesn't produce a notice that's fine, but if it crashes then you do need to check for hasStartTime and hasEndTime.

Reply:
If these values are `null`, a `MissingRequireValueFieldError` is generated and added to the notice container, plus the GtfsEntity is not added to its `GtfsTableContainer`.
So, when executing `validate()` it is assumed that `hasStarTime()`  and `hasEndTime()` will return `true` if these methods are called.
I can add details to the comment to clarify this if needed.

I think this test cover such situations: https://github.com/MobilityData/gtfs-validator/blob/89d069a0a5022bf950c058c9f070e2aaacda02bf/main/src/test/java/org/mobilitydata/gtfsvalidator/table/GtfsFrequencyTableLoaderTest.java#L66

Reply:
>I can add details to the comment to clarify this if needed.

Yes, that would help. If I'm understanding correctly, maybe this would be better:

>validate() will only be called if startTime and endTime have been populated for this frequency